### ISSUE TITLE
Model Wishlist

### ISSUE THREAD

**Issue by EricLBuehler** [EricLBuehler/mistral.rs#156]
Please let us know what model architectures you would like to be added!

- [x] mistralai/Mistral-7B-Instruct-v0.1
- [x] mistralai/Mistral-7B-Instruct-v0.2
- [x] mistralai/Mixtral-8x7B-Instruct-v0.1
- [x] meta-llama/Llama-2-13b-hf
- [x] google/gemma-7b-it
- [x] microsoft/phi-2
- [ ] stabilityai/stablelm-2-1_6b
- [ ] 01-ai/Yi-6B
- [ ] RWKV/rwkv-6-world-1b6 and RWKV/rwkv-5-world-1b5
- [ ] yzsydlc/qwen2
- [ ] adept/persimmon-8b-chat
- [ ] mosaicml/mpt-7b

Quantized architectures:
- [x] llama
- [ ] phi

**Comment by NiuBlibing** [comment]
qwen1.5-72B-Chat

**Comment by NiuBlibing** [comment]
llama3

**Comment by EricLBuehler** [comment]
@NiuBlibing, we have llama3 support ready: the README has a few examples. I will add Qwen support shortly.

**Comment by EricLBuehler** [comment]
@NiuBlibing, I just added Qwen2 support. Quantized Qwen2 support will be added in the next few days.

**Comment by cargecla1** [comment]
Can you add https://huggingface.co/Snowflake/snowflake-arctic-instruct?

**Comment by francis2tm** [comment]
Hello!
Any plans for adding multimodal (e.g. llava) and embedding models?

**Comment by EricLBuehler** [comment]
> Can you add https://huggingface.co/Snowflake/snowflake-arctic-instruct?

@cargecla1, yes! It will be a great use case for ISQ.

**Comment by EricLBuehler** [comment]
> Hello!
> Any plans for adding multimodal (e.g. llava) and embedding models? 

@francis2tm, yes. I plan on supporting Llava and embedding models this week.

**Comment by EricLBuehler** [comment]
@NiuBlibing, you can run Qwen now with ISQ, which will quantize it.

**Comment by kir-gadjello** [comment]
Would be nice to support at least one strong vision-language model: https://huggingface.co/openbmb/MiniCPM-V-2 https://huggingface.co/OpenGVLab/InternVL-Chat-V1-5 with an option to compute visual frontend model on CPU. You might find it easier to ship visual transformer part via onnx.

**Comment by chelbos** [comment]
Would love to see some DeepSeek-VL, this model is better than Llava and spupports multiple images per prompt
https://huggingface.co/collections/deepseek-ai/deepseek-vl-65f295948133d9cf92b706d3

**Comment by chelbos** [comment]
Also, outside the LLM world, would love to see support for https://github.com/cvg/LightGlue :) but not sure if that's possible ...

**Comment by jett06** [comment]
Could you add support to for GGUF quantized Phi-3 to the wishlist? Currently, this fails (built from master):
```sh
Running `./mistralrs-server gguf -m PrunaAI/Phi-3-mini-128k-instruct-GGUF-Imatrix-smashed -t microsoft/Phi-3-mini-128k-instruct -f /home/jett/Downloads/llms/Phi-3-mini-128k-instruct-q3_K_S.gguf`
2024-04-29T03:08:35.180939Z  INFO mistralrs_server: avx: true, neon: false, simd128: false, f16c: false
2024-04-29T03:08:35.180975Z  INFO mistralrs_server: Sampling method: penalties -> temperature -> topk -> topp -> multinomial
2024-04-29T03:08:35.180982Z  INFO mistralrs_server: Loading model `microsoft/Phi-3-mini-128k-instruct` on Cpu...
2024-04-29T03:08:35.180989Z  INFO mistralrs_server: Model kind is: quantized from gguf (no adapters)
2024-04-29T03:08:35.181017Z  INFO hf_hub: Token file not found "/home/jett/.cache/huggingface/token"    
2024-04-29T03:08:35.181048Z  INFO mistralrs_core::utils::tokens: Could not load token at "/home/jett/.cache/huggingface/token", using no HF token.
2024-04-29T03:08:35.181122Z  INFO hf_hub: Token file not found "/home/jett/.cache/huggingface/token"    
2024-04-29T03:08:35.181133Z  INFO mistralrs_core::utils::tokens: Could not load token at "/home/jett/.cache/huggingface/token", using no HF token.
Error: Unknown GGUF architecture `phi3`
```

**Comment by rodion-m** [comment]
It'll be great to see WizardLM-2 and suzume. And thanks for a great tool!

**Comment by W4G1** [comment]
Command-R and Command-R+ from Cohere would be amazing 🙏

**Comment by EricLBuehler** [comment]
@kir-gadjello 
> Would be nice to support at least one strong vision-language model: https://huggingface.co/openbmb/MiniCPM-V-2 https://huggingface.co/OpenGVLab/InternVL-Chat-V1-5 with an option to compute visual frontend model on CPU. You might find it easier to ship visual transformer part via onnx.

Supporting a vision+language or multimodal model is very high priority right now.

---

@chelbos 
> Would love to see some DeepSeek-VL, this model is better than Llava and spupports multiple images per prompt
https://huggingface.co/collections/deepseek-ai/deepseek-vl-65f295948133d9cf92b706d3

I'll add this one too.

> Also, outside the LLM world, would love to see support for https://github.com/cvg/LightGlue :) but not sure if that's possible ...

I will look into it!

---
@jett06 
> Could you add support to for GGUF quantized Phi-3-Mini to the wishlist?

Yes, absolutely, I think it should be easy. In the meantime, you can use ISQ to get the same speed.

---
@rodion-m 
> It'll be great to see WizardLM-2 and suzume. And thanks for a great tool!

Thanks! I think suzume is just finetuned Llama so that can be used already. I'll add WizardLM.

---
@W4G1 
> Command-R and Command-R+ from Cohere would be amazing 🙏

Yes, I'll add those.

---
@yongkangzhao
> T5 and LLaVA

Yes, I'll add those. T5 will be a nice smaller model.

**Comment by jett06** [comment]
@EricLBuehler Thanks for your reply, for adding my suggestion to the model wishlist, and for developing such an awesome project! It's very appreciated :)

**Comment by ldt** [comment]
Congrats for your great work!
+1 for vision models like Idefics2-8b or better would be awesome

**Comment by progressionnetwork** [comment]
Hello, first of all, I want to express my appreciation for the excellent work your team has accomplished on the mistral.rs engine. It's a great project.

I am currently developing a personal AI assistant using Rust, and I believe integrating additional features into your engine could significantly enhance its utility and appeal. Specifically, adding support for Whisper and incorporating Text-to-Speech (TTS) functionalities, such as StyleTTS or similar technologies, would be incredibly beneficial. This would enable the engine to handle LLM inference, speech-to-text, and text-to-speech processes in a unified system very fast (near runtime).

Implementing these features could transform the engine into a more versatile tool for developers like myself, who are keen on building more integrated and efficient AI applications.

**Comment by EricLBuehler** [comment]
@jett06, I just added quantized GGUF Phi-3 support in #276! That is without LongRope support currently, but you can use a `plain` model with ISQ.

**Comment by jett06** [comment]
@EricLBuehler Woah, thank you so much! This will be lovely for us folks with less powerful computers or size constraints, you're awesome :)

**Comment by EricLBuehler** [comment]
@jett06, my pleasure! I just fixed a small bug (in case you saw the strange behavior), so it should be all ready to go now!

**Comment by NeroHin** [comment]
IBM's Granite series Code Models.

[Granite Code Models](https://huggingface.co/collections/ibm-granite/granite-code-models-6624c5cec322e4c148c8b330)

**Comment by LLukas22** [comment]
@NeroHin 



> IBM's Granite series Code Models.
> 
> [Granite Code Models](https://huggingface.co/collections/ibm-granite/granite-code-models-6624c5cec322e4c148c8b330)

The `3b` and `8b` variants should already be supported as they are just based on the `llama` architecture. 

The `20b` and `34b` variants are based on the `GPTBigCode` architecture which currently isn't implemented in `mistral.rs`.

**Comment by chenwanqq** [comment]
> Hello! Any plans for adding multimodal (e.g. llava) and embedding models?

I'm working on it now.[chenwanqq/candle-llava](https://github.com/chenwanqq/candle-llava)
It's not easy dude, tons of image preprocess and tensor concat.

**Comment by EricLBuehler** [comment]
> I'm working on it now.[chenwanqq/candle-llava](https://github.com/chenwanqq/candle-llava)
It's not easy dude, tons of image preprocess and tensor concat.

Yes! Would you be willing to contribute your implementation here once it's done?

**Comment by chenwanqq** [comment]
> > I'm working on it now.[chenwanqq/candle-llava](https://github.com/chenwanqq/candle-llava)
> > It's not easy dude, tons of image preprocess and tensor concat.
> 
> Yes! Would you be willing to contribute your implementation here once it's done?

Yes of course!

**Comment by EricLBuehler** [comment]
> Yes of course!

Great, looking forward to that! I'm working on Idefics 2 here: #309.

**Comment by songyaheng** [comment]
YI-34B

**Comment by niranjanakella** [comment]
Google T5 would be a really cool addition.

**Comment by bachp** [comment]
https://huggingface.co/microsoft/Phi-3-vision-128k-instruct would be nice.

**Comment by EricLBuehler** [comment]
> https://huggingface.co/microsoft/Phi-3-vision-128k-instruct would be nice.

@bachp, #309 is working on vision support. It adds a lot of infrastructure for that, so when I merge it I should be able to add that model soon1

**Comment by chenwanqq** [comment]
> > Yes of course!
> 
> Great, looking forward to that! I'm working on Idefics 2 here: #309.

I have made significant progress in the candle-llava project. However, I hope you don't mind me contributing the code to the official candle project first, and then adapting some infrastructure changes in [#309](https://github.com/EricLBuehler/mistral.rs/pull/309 ) when it have been settled. In addition, I have absolutely no experience with quantization or LoRA, so it may take some time for me to learn.

**Comment by nicarq** [comment]
For embedding models it would be awesome to support snowflake-arctic-embed 🙏

**Comment by EricLBuehler** [comment]
> For embedding models it would be awesome to support snowflake-arctic-embed 🙏

@nicarq, sure! I'll try to do it this weekend.

> I have made significant progress in the candle-llava project. However, I hope you don't mind me contributing the code to the official candle project first, and then adapting some infrastructure changes in https://github.com/EricLBuehler/mistral.rs/pull/309 when it have been settled. In addition, I have absolutely no experience with quantization or LoRA, so it may take some time for me to learn.

@chenwanqq that's fine! Actually, after I merge #309 or #351 (I'll let you know if that's all right), I think it would be enough for you to open a PR where you add your code without LoRA, quantization, device mapping, etc. Basically, you would do (hopefully minimal) porting work, and then I can add those features later. Does that sound good?

**Comment by chenwanqq** [comment]
> > For embedding models it would be awesome to support snowflake-arctic-embed 🙏
> 
> @nicarq, sure! I'll try to do it this weekend.
> 
> > I have made significant progress in the candle-llava project. However, I hope you don't mind me contributing the code to the official candle project first, and then adapting some infrastructure changes in #309 when it have been settled. In addition, I have absolutely no experience with quantization or LoRA, so it may take some time for me to learn.
> 
> @chenwanqq that's fine! Actually, after I merge #309 or #351 (I'll let you know if that's all right), I think it would be enough for you to open a PR where you add your code without LoRA, quantization, device mapping, etc. Basically, you would do (hopefully minimal) porting work, and then I can add those features later. Does that sound good?

Great! I'll do my work!

**Comment by EricLBuehler** [comment]
Hi @chenwanqq! I noticed that your Candle PR was merged - congratulations! Would you have the time to contribute your Llava model here? If you do, I can add the LoRA/quantization/device mapping features in a later PR.

**Comment by solaoi** [comment]
I would also like to see support for quantized models using the Importance Matrix below.
https://github.com/ggerganov/llama.cpp/pull/4861

**Comment by chenwanqq** [comment]
> Hi @chenwanqq! I noticed that your Candle PR was merged - congratulations! Would you have the time to contribute your Llava model here? If you do, I can add the LoRA/quantization/device mapping features in a later PR.

No problem, just waiting for #309 ( I hope API of vision model can be as align as possible).

**Comment by EricLBuehler** [comment]
> No problem, just waiting for https://github.com/EricLBuehler/mistral.rs/pull/309 ( I hope API of vision model can be as align as possible).

Would a merge of #351 also work for that? The 2 PRs use the same API.

**Comment by chenwanqq** [comment]
> > No problem, just waiting for #309 ( I hope API of vision model can be as align as possible).
> 
> Would a merge of #351 also work for that? The 2 PRs use the same API.

I think I will start this part of the work after this weekend. I spent some time the other day studying how to implement the GPU version of nonzero, but I encountered some [difficulties.](https://github.com/huggingface/candle/issues/2258) I think I will come back here before I figure it out.

**Comment by EricLBuehler** [comment]
Hi @chenwanqq I think that using the CPU-only approach for now should be fine. Please see our API: https://github.com/EricLBuehler/mistral.rs/blob/e4a4947db3279e29c4d9544e55f61c1ceaed96b9/mistralrs-core/src/layers.rs#L588-L616

**Comment by siriux** [comment]
Is Phi-3 Medium supported?

I get this error when I send the first message in interactive mode:

```
2024-06-08T18:58:31.530887Z ERROR mistralrs_core::engine: prompt step - Model failed with error: ShapeMismatchBinaryOp { lhs: [1, 16, 1280], rhs: [1, 16, 40, 128], op: "reshape" }
2024-06-08T18:58:31.531072Z ERROR mistralrs_server::interactive_mode: Got a model error: "shape mismatch in reshape, lhs: [1, 16, 1280], rhs: [1, 16, 40, 128]", response: ChatCompletionResponse { id: "0", choices: [Choice { finish_reason: "error", index: 0, message: ResponseMessage { content: "", role: "assistant" }, logprobs: None }], created: 1717873111, model: ".", system_fingerprint: "local", object: "chat.completion", usage: Usage { completion_tokens: 0, prompt_tokens: 16, total_tokens: 16, avg_tok_per_sec: 134.45378, avg_prompt_tok_per_sec: inf, avg_compl_tok_per_sec: NaN, total_time_sec: 0.119, total_prompt_time_sec: 0.0, total_completion_time_sec: 0.0 } }
```

Thanks

**Comment by EricLBuehler** [comment]
Hi @siriux, can you please raise an issue to track progress with the command you ran with `RUST_BACKTRACE=1`? Thanks you!

**Comment by siriux** [comment]
Hi @EricLBuehler, I'm running it with a local gguf, not using huggingface, maybe that's the difference.

I've opened this issue with more details: https://github.com/EricLBuehler/mistral.rs/issues/413

Thanks

**Comment by chenwanqq** [comment]
> Hi @chenwanqq I think that using the CPU-only approach for now should be fine. Please see our API:
> 
> https://github.com/EricLBuehler/mistral.rs/blob/e4a4947db3279e29c4d9544e55f61c1ceaed96b9/mistralrs-core/src/layers.rs#L588-L616

I managed to write a GPU version of nonzero [candle-nonzero](https://github.com/chenwanqq/candle-nonzero).
Actually according to my benchmark test, rayon has too many context cost and it can't even compare with vanilla sequential computing.

**Comment by EricLBuehler** [comment]
That is great! Would it be OK if I add this into mistral.rs, or would you like to contribute it?

**Comment by chenwanqq** [comment]
> That is great! Would it be OK if I add this into mistral.rs, or would you like to contribute it?

I'll try to do it.😁
Since it will introduce some Cuda codes, the building process may be influenced. So please check it after I start a pull request.
(❁´◡`❁)

**Comment by EricLBuehler** [comment]
Ok, great.

**Comment by EricLBuehler** [comment]
Not a problem 😄

**Comment by wseaton** [comment]
> @NeroHin
> 
> > IBM's Granite series Code Models.
> > [Granite Code Models](https://huggingface.co/collections/ibm-granite/granite-code-models-6624c5cec322e4c148c8b330)
> 
> The `3b` and `8b` variants should already be supported as they are just based on the `llama` architecture.
> 
> The `20b` and `34b` variants are based on the `GPTBigCode` architecture which currently isn't implemented in `mistral.rs`.

The 3b and 8b variants do not work out of the box, they rely on tie word embeddings (which I was able to get working), but the BPE tokenizer breaks because there are some tokens in the vocab list that are > 255 characters.

+1 to getting support for GPTBigCode and other starcoder model variants.

**Comment by chenwanqq** [comment]
@EricLBuehler  I'm stil working on LLaVA. Meanwhile, with so much experience with rust and Candle, have you ever encountered any problem about memory usage? I have some kinds of confusion. https://github.com/huggingface/candle/issues/2273#issue-2360380212

**Comment by EricLBuehler** [comment]
@chenwanqq, that is great, let me know if I can help!

I replied to the discussion 2272. However, I discovered that the shadowing does mean that the big tensor will not get dropped! See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b4cd06ea640cefef422b261787ff38e3) and [my comment](https://github.com/huggingface/candle/discussions/2272#discussioncomment-9809946) for more details.

I'll add a clippy lint here to avoid this on our end.

**Comment by bachp** [comment]
@EricLBuehler What is missing for [GGUF quantized Qwen2](https://huggingface.co/Qwen/Qwen2-0.5B-Instruct-GGUF)?

**Comment by EricLBuehler** [comment]
Hi @bachp, that should be relatively easy to add, it would take inspiration from the other GGUF models such as `quantized_phi3.rs`. Do you think you would be able to add this?

**Comment by EricLBuehler** [comment]
We will be adding the Gemma 2 models shortly, see #486!

**Comment by EricLBuehler** [comment]
@francis2tm  @chelbos  @yongkangzhao we just merged LLaVA and LLaVA Next support. Kudos to @chenwanqq for their great work!

For vision models we now have:
- Idefics 2 
- Phi 3 vision
- LLaVA and LLaVA Next

**Comment by csicar** [comment]
I may be able to provide an implementation for whisper sst. If there is interest in that

**Comment by sammcj** [comment]
It doesn't look like it's been mentioned yet but DeepSeek Coder v2 (lite) support would be amazing given it's probably the best coding model out there.

**Comment by EricLBuehler** [comment]
@csicar that would be amazing!

**Comment by EricLBuehler** [comment]
> It doesn't look like it's been mentioned yet but DeepSeek Coder v2 (lite) support would be amazing given it's probably the best coding model out there.

@sammcj that would be great, I can add that.

**Comment by bachp** [comment]
> Hi @bachp, that should be relatively easy to add, it would take inspiration from the other GGUF models such as `quantized_phi3.rs`. Do you think you would be able to add this?

Not sure I'm up to the task yet. However I noticed that candle added support for [quantized Qwen2](https://github.com/huggingface/candle/commit/c63048d3748649c6f13148eb01e6d812d897a0d2), can we re-use this?

**Comment by EricLBuehler** [comment]
@bachp yes. If you want to add that, feel free. I can take a look in a few days.

**Comment by joshpopelka20** [comment]
I'd like to try mistralai/Mistral-Nemo-Instruct-2407 [https://huggingface.co/mistralai/Mistral-Nemo-Instruct-2407](https://huggingface.co/mistralai/Mistral-Nemo-Instruct-2407.) Sounds like it has a similar architecture to Mistral 7B so hoping it won't be too much work.

**Comment by EricLBuehler** [comment]
@joshpopelka20 I just merged this in #595!

```
cargo run --release --features ... -- -i --isq Q4K plain -m mistralai/Mistral-Nemo-Instruct-2407 -a mistral
```

**Comment by joshpopelka20** [comment]
@EricLBuehler thanks for adding that feature. I haven't been able to get it to run as I'm having an issue with paged attention code. I'll add an issue to track and give me details.

**Comment by fredconex** [comment]
Codestral Mamba

**Comment by Remember20240719** [comment]
Hello, thank you for open-sourcing this project!

I would be interested in running Mistral Large Instruct 2407 GGUF.

Trying to run inference on the [Q5 K S quant](https://huggingface.co/bartowski/Mistral-Large-Instruct-2407-GGUF/tree/main/Mistral-Large-Instruct-2407-Q5_K_S) with mistral.rs commit 38fb942 I get : 

```shell
export MISTRALRS_DEBUG=1 ./target/release/./mistralrs-server --port 1234 --throughput gguf --quantized-model-id $D/models/ --quantized-filename Mistral-Large-Instruct-2407-Q5_K_S-00001-of-00003.gguf 
2024-07-28T17:47:45.176615Z  INFO mistralrs_server: avx: true, neon: false, simd128: false, f16c: true
2024-07-28T17:47:45.176929Z  INFO mistralrs_server: Sampling method: penalties -> temperature -> topk -> topp -> minp -> multinomial
2024-07-28T17:47:45.177583Z  INFO mistralrs_server: Model kind is: quantized from gguf (no adapters)
2024-07-28T17:47:45.180482Z  INFO mistralrs_core::pipeline::paths: Loading `Mistral-Large-Instruct-2407-Q5_K_S-00001-of-00003.gguf` locally at `$D/models/Mistral-Large-Instruct-2407-Q5_K_S-00001-of-00003.gguf`
2024-07-28T17:47:45.181595Z  INFO mistralrs_core::pipeline::gguf: Loading model `$D/models/` on cpu.
2024-07-28T17:47:45.266615Z  INFO mistralrs_core::pipeline::gguf: Model config:
general.architecture: llama
general.basename: Mistral
general.file_type: 16
general.finetune: Instruct
general.languages: en, fr, de, es, it, pt, zh, ja, ru, ko
general.license: other
general.license.link: https://mistral.ai/licenses/MRL-0.1.md
general.license.name: mrl
general.name: Mistral Large Instruct 2407
general.quantization_version: 2
general.size_label: Large
general.type: model
general.version: 2407
llama.attention.head_count: 96
llama.attention.head_count_kv: 8
llama.attention.layer_norm_rms_epsilon: 0.00001
llama.block_count: 88
llama.context_length: 131072
llama.embedding_length: 12288
llama.feed_forward_length: 28672
llama.rope.dimension_count: 128
llama.rope.freq_base: 1000000
llama.vocab_size: 32768
quantize.imatrix.chunks_count: 148
quantize.imatrix.dataset: /training_dir/calibration_datav3.txt
quantize.imatrix.entries_count: 616
quantize.imatrix.file: /models_out/Mistral-Large-Instruct-2407-GGUF/Mistral-Large-Instruct-2407.imatrix
split.count: 3
split.no: 0
split.tensors.count: 795
2024-07-28T17:47:45.267503Z  INFO mistralrs_core::pipeline::gguf: Debug is enabled, wrote the names and information about each tensor to `mistralrs_gguf_tensors.txt`.
2024-07-28T17:47:45.316860Z  INFO mistralrs_core::gguf::gguf_tokenizer: GGUF tokenizer model is `llama`, kind: `Unigram`, num tokens: 32768, num added tokens: 0, num merges: 0, num scores: 32768
2024-07-28T17:47:45.316880Z  INFO mistralrs_core::gguf::gguf_tokenizer: Tokenizer: Tokenizer(TokenizerImpl { normalizer: Some(Sequence(Sequence { normalizers: [Prepend(Prepend { prepend: "▁" }), Replace(Replace { pattern: String(" "), content: "▁", regex: SysRegex { regex: Regex { raw: 0x571b958c9200 } } })] })), pre_tokenizer: None, model: Unigram(Unigram { vocab: 32768, unk_id: Some(0), byte_fallback: true }), post_processor: None, decoder: Some(Sequence(Sequence { decoders: [Replace(Replace { pattern: String("▁"), content: " ", regex: SysRegex { regex: Regex { raw: 0x571b958c9000 } } }), ByteFallback(ByteFallback { type_: MustBe!("ByteFallback") }), Fuse(Fuse { type_: MustBe!("Fuse") }), Strip(Strip { content: ' ', start: 1, stop: 0 })] })), added_vocabulary: AddedVocabulary { added_tokens_map: {"</s>": 2, "<unk>": 0, "<s>": 1}, added_tokens_map_r: {2: AddedToken { content: "</s>", single_word: false, lstrip: false, rstrip: false, normalized: false, special: true }, 0: AddedToken { content: "<unk>", single_word: false, lstrip: false, rstrip: false, normalized: false, special: true }, 1: AddedToken { content: "<s>", single_word: false, lstrip: false, rstrip: false, normalized: false, special: true }}, added_tokens: [], special_tokens: [AddedToken { content: "<s>", single_word: false, lstrip: false, rstrip: false, normalized: false, special: true }, AddedToken { content: "</s>", single_word: false, lstrip: false, rstrip: false, normalized: false, special: true }, AddedToken { content: "<unk>", single_word: false, lstrip: false, rstrip: false, normalized: false, special: true }], special_tokens_set: {"<unk>", "<s>", "</s>"}, split_trie: (AhoCorasick(dfa::DFA(
D 000000: \x00-\x0E => 0
F 000016:
* 000032: \x00-\x0E => 0
 matches: 1
* 000048: \x00-\x0E => 0
 matches: 2
* 000064: \x00-\x0E => 0
 matches: 0
 >000080: \x00-\x02 => 80, \x03 => 208, \x04-\x0E => 80
  000096: \x00-\x02 => 0, \x03 => 208, \x04-\x0E => 0
  000112: \x00-\x02 => 80, \x03 => 208, \x04-\n => 80, \x0B => 128, \x0C-\x0E => 80
  000128: \x00-\x02 => 80, \x03 => 208, \x04 => 80, \x05 => 32, \x06-\x0E => 80
  000144: \x00-\x02 => 80, \x03 => 208, \x04 => 80, \x05 => 64, \x06-\x0E => 80
  000160: \x00-\x02 => 80, \x03 => 208, \x04-\x08 => 80, \t => 176, \n-\x0E => 80
  000176: \x00-\x02 => 80, \x03 => 208, \x04-\x06 => 80, \x07 => 192, \x08-\x0E => 80
  000192: \x00-\x02 => 80, \x03 => 208, \x04 => 80, \x05 => 48, \x06-\x0E => 80
  000208: \x00 => 80, \x01 => 112, \x02 => 80, \x03 => 208, \x04-\n => 80, \x0B => 144, \x0C => 80, \r => 160, \x0E => 80
match kind: LeftmostLongest
prefilter: true
state length: 14
pattern length: 3
shortest pattern length: 3
longest pattern length: 5
alphabet length: 15
stride: 16
byte classes: ByteClasses(0 => [0-46], 1 => [47], 2 => [48-59], 3 => [60], 4 => [61], 5 => [62], 6 => [63-106], 7 => [107], 8 => [108-109], 9 => [110], 10 => [111-114], 11 => [115], 12 => [116], 13 => [117], 14 => [118-255])
memory usage: 992
)
), [1, 2, 0]), split_normalized_trie: (AhoCorasick(dfa::DFA(
D 000000: \x00 => 0
F 000001:
 >000002: \x00 => 2
  000003: \x00 => 0
match kind: LeftmostLongest
prefilter: false
state length: 4
pattern length: 0
shortest pattern length: 18446744073709551615
longest pattern length: 0
alphabet length: 1
stride: 1
byte classes: ByteClasses(0 => [0-255])
memory usage: 16
)
), []), encode_special_tokens: false }, truncation: None, padding: None })
2024-07-28T17:47:45.318706Z  INFO mistralrs_core::gguf::chat_template: Discovered and using GGUF chat template: `{%- if messages[0]['role'] == 'system' %}\n    {%- set system_message = messages[0]['content'] %}\n    {%- set loop_messages = messages[1:] %}\n{%- else %}\n    {%- set loop_messages = messages %}\n{%- endif %}\n\n{{- bos_token }}\n{%- for message in loop_messages %}\n    {%- if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}\n        {{- raise_exception('After the optional system message, conversation roles must alternate user/assistant/user/assistant/...') }}\n    {%- endif %}\n    {%- if message['role'] == 'user' %}\n        {%- if loop.last and system_message is defined %}\n            {{- '[INST] ' + system_message + '\n\n' + message['content'] + '[/INST]' }}\n        {%- else %}\n            {{- '[INST] ' + message['content'] + '[/INST]' }}\n        {%- endif %}\n    {%- elif message['role'] == 'assistant' %}\n        {{- ' ' + message['content'] + eos_token}}\n    {%- else %}\n        {{- raise_exception('Only user and assistant roles are supported, with the exception of an initial optional system message!') }}\n    {%- endif %}\n{%- endfor %}\n`
Error: cannot find tensor info for output_norm.weight
```

Attached: [mistralrs_gguf_tensors.txt](https://github.com/user-attachments/files/16404739/mistralrs_gguf_tensors.txt)

**Comment by dancixx** [comment]
Hi guys, thanks for the awesome work. Is there any plan to support Idefics3 and InternVl2?

**Comment by bhupesh-sf** [comment]
hey, thanks for this awesome work as it allows people with fewer resources to run LLMs and VLMs on their machines.

Are we planning to support TTS, STT and image generation models as well? There is a lot of buzz around Flux.1 these days. There are also some good open-source models out there for voice cloning etc.

But once again I must appreciate projects like these to help out the community. 🥇

**Comment by EricLBuehler** [comment]
@bhupesh-sf, yes, I'm planning to expand into the multimodal space with a broad variety of models. As you suggested, TTS, STT, and image generation are all on the table as well as embedding models.

@dancixx, yes, I plan to add Idefics 3 at least!

**Comment by jasinco** [comment]
So does it support Deepseek Coder yet?

**Comment by pigfoot** [comment]
I'd appreciate if Qwen2-VL could be considered to add: https://huggingface.co/Qwen/Qwen2-VL-2B-Instruct

**Comment by ariqpradipa** [comment]
I suggest considering the addition of `https://huggingface.co/openbmb/MiniCPM3-4B` as well.

**Comment by ethanc8** [comment]
Here are two open image+text to image+text models (these are the only ones I know of):

Anole - Chameleon finetune with image output
[Code](https://github.com/GAIR-NLP/anole) | [arXiv](https://arxiv.org/pdf/2407.06135) | [7B weights](https://huggingface.co/GAIR/Anole-7b-v0.1/tree/main/models/7b) | [Website](https://gair-nlp.github.io/anole/)

Lumina-mGPT - an independent project by Chinese researchers
[training/inference code](https://github.com/Alpha-VLLM/Lumina-mGPT) | [paper](https://arxiv.org/pdf/2408.02657) | [hf](https://huggingface.co/collections/Alpha-VLLM/lumina-mgpt-family-66ae48a59a973eeae4513848)

The Lumina project also made many text-to-other-modes models, in their [Lumina-T2X](https://github.com/Alpha-VLLM/Lumina-T2X) subproject.

**Comment by EricLBuehler** [comment]
Quick status update:

We have added the FLUX and Llama 3.2 Vision models recently, with Parler TTS support coming in #791.

If anyone would be able to attempt an implementation of any of the requested models, that would be incredible!

My idea of the priority of adding models is:
1) Parler TTS
2) Idefics 3
3) Pixtral
4) QwenVL-2
5) Anole/Chameleon

Please feel free to revise this order!

@youcefs21 I think Pixtral would certainly be an interesting addition! I will take a look at adding that.

@ethanc8 thanks for linking the models! I think Chameleon would be a cool add, the Anole model seems very interesting.

@pigfoot We could add QwenVL-2 :)

@dancixx Idefics 3 support was recently merged into transformers, so work can begin on that too.

@bhupesh-sf I have added FLUX with Parler TTS being implemented in #791!

**Comment by EricLBuehler** [comment]
@oldgithubman I just merged support for Qwen 2.5 in #805!

**Comment by ChristianWeyer** [comment]
> My idea of the priority of adding models is:
> 
> 1. Parler TTS
> 2. Idefics 3
> 3. Pixtral
> 4. QwenVL-2
> 5. Anole/Chameleon

Move Qwen2 VL up the list @EricLBuehler. It is super strong and super important - especially with the HTTP API server (and improved Metal support ;-)).

**Comment by EricLBuehler** [comment]
@ChristianWeyer that sounds good :)

**Comment by bhupesh-sf** [comment]
Anole seems quite interesting, would prioritize it over pixtral as it can generate images as well

**Comment by ethanc8** [comment]
Another interesting highly multimodal model is [Emu3-Gen](https://huggingface.co/BAAI/Emu3-Gen), which can take in images, video, and text, and output images, video, and text -- its video generation is slightly better than OpenSora, and it's also able to extend existing videos. You can see example generations on [the website](https://emu.baai.ac.cn/about).

**Comment by xydz** [comment]
Please support the glm-4-9b-chat model. The model address link is https://huggingface.co/THUDM/glm-4-9b-chat.

**Comment by mseri** [comment]
What about zamba2 [2.7B](https://huggingface.co/Zyphra/Zamba2-2.7B) and [7B}(https://huggingface.co/Zyphra/Zamba2-7B) by [Zyphra](https://huggingface.co/Zyphra/)? They should be quite decent small models to try out

**Comment by cjs-axsh** [comment]
https://huggingface.co/jinaai/jina-embeddings-v3

**Comment by czonios** [comment]
Is there any way this library would support Stable Diffusion in the near future (as I saw that FLUX is already supported) with quantization and LoRA adapter capabilities?

**Comment by GraphicalDot** [comment]
support for black-forest-labs/FLUX.1-Fill-dev .uqff format.

**Comment by hak8or** [comment]
The new qwq model that came out a week or so ago. It's very well regarded on the locallama subreddit (seems to be the communities darling at the moment), and tinkering with it myself has also yielded positive results. To be able to use it in mistral.rs would be fantastic!

https://huggingface.co/Qwen/QwQ-32B-Preview

**Comment by dancixx** [comment]
Hi guys,

is there a plan to support llama 3.3?

**Comment by EricLBuehler** [comment]
Hi @dancixx , Llama 3.3 is already supported as it uses the same architecture are 3.1/3.2.

**Comment by EricLBuehler** [comment]
Hi @dancixx , Llama 3.3 is already supported as it uses the same architecture are 3.1/3.2.

**Comment by franklucky001** [comment]
support deepseek language model v2 v3, please. [https://huggingface.co/deepseek-ai/DeepSeek-V3](url)

**Comment by EricLBuehler** [comment]
@franklucky001 I just merged #1010 which adds Deepseek v2!

**Comment by EricLBuehler** [comment]
@test3211234 

> Also, if I wanted to share this, how would I go about it? Do I just put this in the Cargo.toml:

Yes, you would add `mistralrs = { git = "https://github.com/EricLBuehler/mistral.rs.git" }`.

> can you recommend me a model for video text detection?

I'd check out something like [this LLaVA model](llava-hf/LLaVA-NeXT-Video-7B-hf) or [Qwen2-VL](Qwen/Qwen2-VL-7B-Instruct) if you want to strictly take video input, otherwise, using any vision model frame-by-frame (or skipping some using some heuristic) might work too!

**Comment by youcefs21** [comment]
@test3211234 if you just need text, why not process frame by frame with something like https://huggingface.co/stepfun-ai/GOT-OCR2_0

should be relatively fast on a decent GPU, like 1-2 seconds per frame, skipping every other frame, you can probably do 45 minutes of video text extraction per day with a naive approach. You can probably optimize that more with some sort of parallelism.

**Comment by andreclaudino** [comment]
Would be great to support Loras for Flux. Any tip or guide on how I can contribute for that?

**Comment by EricLBuehler** [comment]
@test3211234 you can run the distill models without any changes as they are simple Qwen/Llama models. I merged initial support for the DeepSeek R1 model for those with large enough systems and will support loading from the FP8 weights shortly.

**Comment by mush42** [comment]
@EricLBuehler Please add support for `SmolVLM `:
https://huggingface.co/blog/smolervlm

**Comment by EricLBuehler** [comment]
@mush42 you can already run all the SmolVLM models: https://github.com/EricLBuehler/mistral.rs/blob/master/docs/IDEFICS3.md

They share an architecture with Idefics 3, so all you need to do is change the model ID.

**Comment by like-a-freedom** [comment]
Vote for moondream2: https://huggingface.co/vikhyatk/moondream2

**Comment by francis2tm** [comment]
https://huggingface.co/deepseek-ai/Janus-Pro-7B including image gen?

**Comment by Remember20240719** [comment]
Hello, Mistral Small Instruct 2501 please. I've filed issue #1118

**Comment by Aveline67** [comment]
Qwen/QwQ32B

**Comment by rozgo** [comment]
> Qwen/QwQ32B

does this work?
```
cargo run --release --features metal -- -i --log test.log --isq Q4K plain -m Qwen/QwQ-32B -a qwen2
```
or cuda instead of metal.

**Comment by green-s** [comment]
Ovis2

https://github.com/AIDC-AI/Ovis
https://huggingface.co/AIDC-AI/Ovis2-34B

It's based on Qwen-2.5 but uses Apple's aim2v for vision, supposedly resulting in better performance than Qwen2.5-VL-72B, and importantly it has a 34B checkpoint while Qwen2.5-VL does not.

**Comment by bet0x** [comment]
@EricLBuehler Can we get support for https://huggingface.co/mistralai/Mistral-Small-3.1-24B-Instruct-2503? I's a day old, Multi Modal.

**Comment by EricLBuehler** [comment]
@bet0x absolutely, working on it now - expect it to be added in the next few days!

**Comment by EricLBuehler** [comment]
@green-s that looks interesting, will check it out & possibly add it after mistral3!

**Comment by bet0x** [comment]
> [@bet0x](https://github.com/bet0x) absolutely, working on it now - expect it to be added in the next few days!

That would be absolutely amazing!

**Comment by EricLBuehler** [comment]
@bet0x PR is up on #1221 - the model mostly implemented already, faster than I expected :)

**Comment by bet0x** [comment]
> [@bet0x](https://github.com/bet0x) PR is up on [#1221](https://github.com/EricLBuehler/mistral.rs/pull/1221) - the model mostly implemented already, faster than I expected :)

Is Quant supported or as is using safetensors?

**Comment by brrr** [comment]
> [@bet0x](https://github.com/bet0x) PR is up on [#1221](https://github.com/EricLBuehler/mistral.rs/pull/1221) - the model mostly implemented already, faster than I expected :)

Awesome! Much faster then I expected too, I was going to ask you @EricLBuehler if you have time to port mistral3.

**Comment by EricLBuehler** [comment]
@bet0x @brrr the Mistral 3 PR #1221 should be runnable now - ISQ is supported!

**Comment by EricLBuehler** [comment]
@bet0x @brrr Mistral 3 support has been merged w/ ISQ support too!

**Comment by bet0x** [comment]
@EricLBuehler I will test it today! thank you for your excellent work !!!

**Comment by di-osc** [comment]
hello @EricLBuehler , thank you for open-sourcing the mistral.rs, it is a fantastic project. can you please support the [qwen2.5-omini](https://huggingface.co/Qwen/Qwen2.5-Omni-7B) model?

**Comment by Atharva-Phatak** [comment]
@EricLBuehler I would love to contribute [gotocr-2.0](https://huggingface.co/stepfun-ai/GOT-OCR2_0). Although I don't know where to start, I would love to do it myself. Looking for some guidance.

**Comment by Korolen** [comment]
GLM-4-32B-0414 / GLM-Z1-32B-0414 / GLM-Z1-Rumination-32B-0414 seem overpowered. would love to see them supported in mistral.rs! big cheers and thanks from me btw for this amazing work / inference setup. <3 !! 🍻  ✊

**Comment by sammcj** [comment]
Qwen3 would be nice!

```
mistralrs-server gguf --quantized-model-id unsloth/Qwen3-0.6B-GGUF --quantized-filename Qwen3-0.6B-UD-Q6_K_XL.gguf
2025-05-01T01:39:30.652604Z  INFO mistralrs_server: avx: false, neon: true, simd128: false, f16c: false
2025-05-01T01:39:30.652623Z  INFO mistralrs_server: Sampling method: penalties -> temperature -> topk -> topp -> minp -> multinomial
2025-05-01T01:39:30.652635Z  INFO mistralrs_server: Model kind is: gguf quantized from gguf (no adapters)
Qwen3-0.6B-UD-Q6_K_XL.gguf [00:00:23] [███████████████████████████████████████████████████████████████████████████████████████████████████████████████] 549.76 MiB/549.76 MiB 25.31 MiB/s (0s)2025-05-01T01:39:56.534520Z  INFO mistralrs_core::pipeline::gguf: Prompt chunk size is 1024.

thread 'main' panicked at mistralrs-core/src/gguf/content.rs:94:22:
called `Result::unwrap()` on an `Err` value: Unknown GGUF architecture `qwen3`
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by EricLBuehler** [comment]
@sammcj absolutely, will add gguf support!

**Comment by anidotnet** [comment]
@EricLBuehler I also want Qwen3 support for GGUF. And also while you are at it, could you please take a look at if it is possible to add all ggml types mentioned here - https://github.com/ggml-org/ggml/blob/17733de6a7854b9696be7a563711c9aa4a34b2d3/include/ggml.h#L351, so that I can use Qwen3 1-bit GGUF model also?

**Comment by Jumaron** [comment]
Gemma-3n 4B and 2B version

**Comment by mattmireles** [comment]
+1 for Gemma-3n!!

**Comment by EricLBuehler** [comment]
@mattmireles @Jumaron Gemma3n support is coming very soon! Check out #1519 for progess. Images and text are already working, audio support is being added.

https://github.com/EricLBuehler/mistral.rs/pull/1519#issuecomment-3026222870

**Comment by tablet0ps** [comment]
Support for [Hunyuan-A13B-Instruct](https://huggingface.co/tencent/Hunyuan-A13B-Instruct) would be very much appreciated! This is one of the models that can achieve quality comparable to flagship models while taking significantly less memory.

**Comment by jac-cbi** [comment]
Please consider adding support for Apple’s recently published DiffuCoder family of models:

https://huggingface.co/collections/apple/diffucoder-6868139f56672ae046fe04e8

**Comment by Revive-Curiosity** [comment]
Please add support for this model https://huggingface.co/THUDM/GLM-4.1V-9B-Thinking . It is a wonderful model for local inference.

**Comment by Revive-Curiosity** [comment]
Please add support for this model https://huggingface.co/baidu/ERNIE-4.5-21B-A3B-PT . It is on par with qwen3 30b A3b but with only 21b total parameters. This will be great model for local inference!

**Comment by Revive-Curiosity** [comment]
Please add support for this model https://huggingface.co/RekaAI/reka-flash-3.1 . It is on par with 03 mini. Another great model for local AI inference.

**Comment by ghost** [comment]
Models by LiquidAI

https://huggingface.co/LiquidAI/collections
https://github.com/Liquid4All/liquid_llama.cpp

**Comment by yunqing0323** [comment]
https://huggingface.co/microsoft/Phi-4-multimodal-instruc

**Comment by KirioXX** [comment]
https://huggingface.co/mistralai/Voxtral-Mini-3B-2507

Would be great as an alternative to all the Whisper tools

**Comment by thanhmt-dev** [comment]
https://huggingface.co/openai/gpt-oss-120b
https://huggingface.co/openai/gpt-oss-20b

Would be great. Thank you.

**Comment by mseri** [comment]
Also the new https://huggingface.co/nvidia/NVIDIA-Nemotron-Nano-9B-v2

**Comment by aminnasiri** [comment]
Automatic Speech Recognition: [openai/whisper](https://huggingface.co/openai/whisper-large-v3-turbo)

**Comment by aminnasiri** [comment]
Google VaultGemma

https://huggingface.co/google/vaultgemma-1b

**Comment by anto-progetica** [comment]
Some models are based on this architecture. Models used for TTS 
CsmForConditionalGeneration
https://huggingface.co/sesame/csm-1b

**Comment by kindywu** [comment]
Deepseek OCR
https://huggingface.co/deepseek-ai/DeepSeek-OCR

**Comment by kindywu** [comment]
MinerU2.5-2509-1.2B
https://huggingface.co/opendatalab/MinerU2.5-2509-1.2B

**Comment by kailasas-auspicious** [comment]
Kimi-k2
https://huggingface.co/moonshotai/Kimi-K2-Instruct

**Comment by kailasas-auspicious** [comment]
moonshotai/Kimi-Audio-7B
https://huggingface.co/moonshotai/Kimi-Audio-7B

**Comment by kailasas-auspicious** [comment]
maya-research/maya1
https://huggingface.co/maya-research/maya1

**Comment by kailasas-auspicious** [comment]
deepseek-ai/Janus-Pro-7B
https://huggingface.co/deepseek-ai/Janus-Pro-7B

**Comment by sempervictus** [comment]
NVIDIA-Nemotron-3-Nano-30B-A3B-BF16 - its a qwen3 with mamba layers

**Comment by jac-cbi** [comment]
Molmo2-8B from allenai:
https://huggingface.co/allenai/Molmo2-8B

```sh
$ ./target/release/mistralrs-server vision-plain -m allenai/Molmo2-8B --hf-cache-path ~/ai_models/
2026-01-02T20:34:49.142042Z  INFO mistralrs_server_core::mistralrs_for_server_builder: avx: false, neon: true, simd128: false, f16c: false
2026-01-02T20:34:49.142081Z  INFO mistralrs_server_core::mistralrs_for_server_builder: Sampling method: penalties -> temperature -> topk -> topp -> minp -> multinomial
2026-01-02T20:34:49.142092Z  INFO mistralrs_server_core::mistralrs_for_server_builder: Model kind is: normal (no adapters)
2026-01-02T20:34:49.142144Z  INFO mistralrs_core::utils::tokens: Could not load token at "/Users/jason/.cache/huggingface/token", using no HF token.
2026-01-02T20:34:49.142220Z  INFO mistralrs_core::pipeline::vision: Loading `tokenizer.json` at `allenai/Molmo2-8B`
2026-01-02T20:34:49.142355Z  INFO mistralrs_core::pipeline::vision: Loading `config.json` at `allenai/Molmo2-8B`
2026-01-02T20:34:49.144180Z  INFO mistralrs_core::pipeline::paths: Read from cache file "/Users/jason/.cache/huggingface/hub/allenai-Molmo2-8B_repo_list.json"
2026-01-02T20:34:49.144300Z  INFO mistralrs_core::pipeline::paths: Found model weight filenames ["model-00001-of-00008.safetensors", "model-00002-of-00008.safetensors", "model-00003-of-00008.safetensors", "model-00004-of-00008.safetensors", "model-00005-of-00008.safetensors", "model-00006-of-00008.safetensors", "model-00007-of-00008.safetensors", "model-00008-of-00008.safetensors"]
# download messages omitted
2026-01-02T20:53:48.620302Z  INFO mistralrs_core::pipeline::vision: Read from cache file "/Users/jason/.cache/huggingface/hub/allenai-Molmo2-8B_repo_list.json"
2026-01-02T20:53:48.620328Z  INFO mistralrs_core::pipeline::vision: Loading `generation_config.json` at `allenai/Molmo2-8B`
generation_config.json [00:00:00] 
# downloaded omitted
2026-01-02T20:53:48.851457Z  INFO mistralrs_core::pipeline::vision: Loading `preprocessor_config.json` at `allenai/Molmo2-8B`
preprocessor_config.json [00:00:00] 
# download omitted
2026-01-02T20:53:49.104266Z  INFO mistralrs_core::pipeline::vision: Loading `processor_config.json` at `allenai/Molmo2-8B`
processor_config.json [00:00:00] 
# download omitted
2026-01-02T20:53:49.332296Z  INFO mistralrs_core::pipeline::vision: Loading `chat_template.jinja` at `allenai/Molmo2-8B`
chat_template.jinja [00:00:00] 
# download omitted
thread 'main' (17430647) panicked at mistralrs-core/src/pipeline/loaders/vision_loaders.rs:338:14:
AutoVisionLoader: Unsupported Hugging Face Transformers -CausalLM model class `Molmo2ForConditionalGeneration`. Please raise an issue.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by vkomenda** [comment]
LiquidAI LFM 2.5
https://huggingface.co/LiquidAI/LFM2.5-1.2B-Instruct

**Comment by qianxiaoming** [comment]
MiniMax-M2.1
https://huggingface.co/MiniMaxAI/MiniMax-M2.1

**Comment by copumpkin** [comment]
Would DINOv3 fit nicely into this framework? I noticed Qwen3 Embedding in the repo, which is somewhat similar in principle.

**Comment by Mo0nbase** [comment]
z.ai GLM-4.7-Flash

https://huggingface.co/zai-org/GLM-4.7-Flash

https://x.com/Zai_org/status/2013261304060866758?s=20

**Comment by venix** [comment]
Qwen released open source speech models, https://qwen.ai/blog?id=qwen3tts-0115 

Any chance  these will be added ?

**Comment by ArEnSc** [comment]
@venix ill probably take it on

**Comment by waqarmalik** [comment]
It'll be great to have support for encoder-decoder model like the [T5Gemma-2](https://huggingface.co/google/t5gemma-2-270m-270m) which is based on Gemma3.

**Comment by curiositry** [comment]
Voxtral Mini 4B Realtime 2602

https://huggingface.co/mistralai/Voxtral-Mini-4B-Realtime-2602

**Comment by magick93** [comment]
Also would like to see support for Voxtral

**Comment by ghost** [comment]
nanbeige4.1 3B

**Comment by SystemicVoid** [comment]
Qwen 3.5 would be wonderful 🙏

**Comment by mjgarton** [comment]
> Qwen 3.5 would be wonderful 🙏

@SystemicVoid  See https://github.com/EricLBuehler/mistral.rs/pull/1956

**Comment by mjgarton** [comment]
> Qwen 3.5 would be wonderful 🙏

Also see https://github.com/EricLBuehler/mistral.rs/pull/1993 .  There are now two PRs open for Qwen 3.5 models.

**Comment by EricLBuehler** [comment]
Thanks for letting me know. @mjgarton and @SystemicVoid, I merged #1993 and we now support Qwen 3.5.

**Comment by ConYel** [comment]
kokoro models are supported?

**Comment by luyui** [comment]
> Thanks for letting me know. [@mjgarton](https://github.com/mjgarton) and [@SystemicVoid](https://github.com/SystemicVoid), I merged [#1993](https://github.com/EricLBuehler/mistral.rs/pull/1993) and we now support Qwen 3.5.

tried to run mlx-community/Qwen3.5-0.8B-MLX-8bit but got:

> Error: shape mismatch for language_model.model.layers.0.linear_attn.in_proj_qkv.weight, expected: [6144, 1024], got: [6144, 256]

are the smaller ones in the series not supported yet?

**Comment by nmb4** [comment]
Rekas Edge 7B Model would be cool, it currently fails with this Error: Unsupported Hugging Face Transformers -CausalLM model class `Yasa2ForConditionalGeneration`

**Comment by rnett** [comment]
Mostly want to use Gemma 4, which seems to already be in progress, but a few others that would be nice:
 * Minimax 2.7 once its weights drop (should be any day now)
 * [DeepSeek 3.2](https://huggingface.co/deepseek-ai/DeepSeek-V3.2) if it isn't compatible with the deepseek 3 support.
 * I probably don't need to mention it but Qwen 3.6 once it drops, too.

**Comment by Orefa** [comment]
BAAI/bge-m3

**Comment by nicolas-geysse** [comment]
+1 for **PaddleOCR-VL-1.5** (`paddleocr_vl` arch) — opened a dedicated issue with details: #2128

Use case: OCR / document understanding pipelines wanting to stay Rust-native instead of falling back to vLLM/TRT-LLM Python sidecars. Multiple weight formats already available on HF (safetensors, GGUF, ONNX, MXFP4 modelopt).

**Comment by sa-** [comment]
Hello and thank you for this awesome project :)

Here are some TTS models that I would like to request

- [OmniVoice](https://huggingface.co/k2-fsa/OmniVoice)
- [Kokoro](https://huggingface.co/hexgrad/Kokoro-82M)
- [VibeVoice-Realtime-0.5B)](https://huggingface.co/microsoft/VibeVoice-Realtime-0.5B) 
- [Qwen3-TTS](https://huggingface.co/Qwen/Qwen3-TTS-12Hz-0.6B-Base)

**Comment by Dakraid** [comment]
Support for IBM's new Granite 4.1 would be very welcomed

**Comment by dvc94ch** [comment]
I can't get the phi-4-multimodal-instruct to work, maybe some more debugging is required on that one? does it apply the additional lora weights? always returns the same answer "The image does not contain any text to perform OCR on."

**Comment by dvc94ch** [comment]
Also Qwen2.5-Omni-3B/7B would be nice

**Comment by thiezn** [comment]
One of the newer image generation models would really close the loop on “having it all”! 

For instance, flux2 or z-image would be great

**Comment by copumpkin** [comment]
Anyone see major obstacles with getting Qwen 3.6 in?

**Comment by EricLBuehler** [comment]
Hey @copumpkin! Qwen 3.6 is already supported as we support the Qwen 3.5 architecture (https://ericlbuehler.github.io/mistral.rs/reference/supported-models/#multimodal-models).

**Comment by misureaudio** [comment]
trying Jackrong/Qwopus3.6-27B-v2

loading is ok

2026-06-29T12:55:10.928505Z  INFO mistralrs_quant::utils::log: DType selected is BF16.
2026-06-29T12:55:11.000382Z  INFO mistralrs_core::pipeline::loaders::auto_device_map: Using automatic device mapping parameters: multimodal[max_seq_len: 4096, max_batch_size: 1, max_image_shape: (1024, 1024), max_num_images: 1].
2026-06-29T12:55:11.000423Z  INFO mistralrs_core::pipeline::loaders::auto_device_map: The following sub-models will not be device mapped and will be loaded on cuda[0]: vision
2026-06-29T12:55:11.000449Z  INFO mistralrs_quant::utils::log: Model has 64 repeating layers.
2026-06-29T12:55:11.000456Z  INFO mistralrs_quant::utils::log: Loading model according to the following repeating layer mappings:
2026-06-29T12:55:11.017128Z  INFO mistralrs_quant::utils::log: Layers 0-63: cuda[0] (92 GB)
2026-06-29T12:55:11.051851Z  INFO mistralrs_quant::utils::log: FlashAttention is enabled.
2026-06-29T12:55:11.051877Z  INFO mistralrs_core::pipeline::multimodal: Loading model weights.
2026-06-29T12:58:16.705546Z  INFO mistralrs_core::paged_attention: Allocating 40176 MB for PagedAttention KV cache per GPU
2026-06-29T12:58:16.705562Z  INFO mistralrs_core::paged_attention: PagedAttention KV cache type is BF16
2026-06-29T12:58:16.705564Z  INFO mistralrs_core::paged_attention: Using PagedAttention with block size 32 and 5022 GPU blocks: available context length is 160704 tokens
2026-06-29T12:58:24.727919Z  INFO mistralrs_server_core::mistralrs_for_server_builder: Model loaded.
2026-06-29T12:58:24.728359Z  INFO mistralrs_core: mistral.rs version: 0.8.23
2026-06-29T12:58:24.728364Z  INFO mistralrs_core: git revision: 15986c037bbe3ee31085d1c73abd2ea3cb11f094
2026-06-29T12:58:24.798670Z  INFO mistralrs_core: Preloaded 735 Candle CUDA PTX functions.
2026-06-29T12:58:24.798818Z  INFO mistralrs_core: Pipeline input modalities are [📝 Text, 🖼️ Vision]
2026-06-29T12:58:24.798823Z  INFO mistralrs_core: Pipeline output modalities are [📝 Text]
2026-06-29T12:58:24.801694Z  INFO mistralrs_core::paged_attention::scheduler: Prefix caching enabled (block-level, PagedAttention). Expect higher multi-turn throughput for both text and multimodal.
2026-06-29T12:58:25.282443Z  INFO mistralrs::commands::run: Model loaded, starting interactive mode...
2026-06-29T12:58:25.282832Z  INFO mistralrs::commands::run::interactive: Starting interactive loop with sampling params: SamplingParams { temperature: Some(0.1), top_k: Some(32), top_p: Some(0.1), min_p: Some(0.05), top_n_logprobs: 0, frequency_penalty: None, presence_penalty: None, repetition_penalty: None, stop_toks: None, max_len: None, logits_bias: None, n_choices: 1, dry_params: Some(DrySamplingParams { sequence_breakers: ["\n", ":", "\"", "*"], multiplier: 0.0, base: 1.75, allowed_length: 2 }) }

but  immediately after the prompt:

Sampling: temp=0.1, top_k=32, top_p=0.1, min_p=0.05
====================
> Does the infinite tetration of i converge in the complex plane?
1FlashInfer decode failed: Error in function 'BatchDecodeWithPagedKVCacheDispatched' at src/cuda/flashinfer/attention/decode.cuh:755: Unsupported group_size: 6
2026-06-29T12:59:42.693717Z  WARN mistralrs_core::pipeline::multimodal: CUDA decode graphs disabled after capture/replay error: FlashInfer decode failed: batch=1 padded_batch=79 qo_heads=24 kv_heads=4 head_size=256
flashinfer_decode failed with status 999
FlashInfer decode failed: Error in function 'BatchDecodeWithPagedKVCacheDispatched' at src/cuda/flashinfer/attention/decode.cuh:755: Unsupported group_size: 6
2026-06-29T12:59:42.695027Z ERROR mistralrs_core::engine: step - Model failed with error: FlashInfer decode failed: batch=1 padded_batch=79 qo_heads=24 kv_heads=4 head_size=256
flashinfer_decode failed with status 999
2026-06-29T12:59:42.695487Z ERROR mistralrs::commands::run::interactive: Got a model error: "FlashInfer decode failed: batch=1 padded_batch=79 qo_heads=24 kv_heads=4 head_size=256\nflashinfer_decode failed with status 999", response: ChatCompletionResponse { id: "1", choices: [Choice { finish_reason: "error", index: 0, message: ResponseMessage { content: Some("1"), role: "assistant", tool_calls: None, reasoning_content: None }, logprobs: None }], created: 1782737982, model: "Jackrong/Qwopus3.6-27B-v2", system_fingerprint: "local", object: "chat.completion", usage: Usage { completion_tokens: 1, prompt_tokens: 23, total_tokens: 24, avg_tok_per_sec: 59.25926, avg_prompt_tok_per_sec: 58.080807, avg_compl_tok_per_sec: 111.111115, total_time_sec: 0.405, total_prompt_time_sec: 0.396, total_completion_time_sec: 0.009 }, agentic_tool_calls: None, files: None, session_id: None }

by the way, identical problem with Qwen/Qwen3.6-27B

**Comment by cornmonger** [comment]
Olmo 3

not sure if it's supported ... i don't see it mentioned in docs

**Comment by haydonryan** [comment]
I looked, but I don't see these yet:
deepseek v4 flash
Minimax M2.5,2.7, M3

While I don't have the hardware to run it yet:
Kimi k2.5, k2.6, k2.7 coder.
GLM 5.2 is all the rage

**Comment by ryaningli** [comment]
Expect to support Qwen3-ASR

**Comment by s3rj1k** [comment]
Would be nice to have this https://huggingface.co/poolside/Laguna-S-2.1 model supported

**Comment by copumpkin** [comment]
How about this shiny new Muse Glimmer? https://huggingface.co/meta-models/Muse-Glimmer-30B

**Comment by haydonryan** [comment]
I'm working on deepseek v4flash additions to candle. just waiting on the new gpu to be delievered so i can do actual testing (not just the stub)

**Comment by lmmx** [comment]
I presume it won’t work out the box (but there is [mention][bonsai] of Bonsai models in the repo!) but the [new Bonsai 2 27B][bonsai2] at q1 GGUF, which requires porting the Hadamard ops and tensor types from their llama.cpp [fork][fork]…

[bonsai]: https://github.com/EricLBuehler/mistral.rs/blob/d5ae0f18f2170f10d30880cb7d21fb0880410e7b/mistralrs-cli/src/commands/quant/gguf_discovery.rs#L947
[bonsai2]: https://prismml.com/news/bonsai-2-27b
[fork]: https://github.com/ggml-org/llama.cpp/compare/master...PrismML-Eng:llama.cpp:prism#diff-8d806e7c33e3df73eed32902b420fbb456ec8aa623b14e29a18081128d636f65
